File: Python/memory_mapping/demo_memory_mapping.py
# ==============================================================================
# demo_memory_mapping.py
# ==============================================================================
#
# Demonstrates the memory-mapping utilities provided by the trex_selector
# package.  Memory-mapped matrices allow handling large datasets without
# loading the entire array into Python's active memory, seamlessly passing
# a pointer to the C++ backend.
#
# Mirrors R/memory_mapping/demo_memory_mapping.R.
#
# ==============================================================================

import os
import tempfile

import numpy as np
from trex_selector.utils import AccessMode, MemoryMappedMatrix, numpy_to_memmap

print("=" * 65)
print(" Memory Mapping Demo for TRexSelector")
print("=" * 65 + "\n")

# ==============================================================================
# 1. Create a sample in-memory matrix
# ==============================================================================

n_rows = 100
n_cols = 50
print(f"1. Creating a sample {n_rows} x {n_cols} numeric matrix.\n")

rng = np.random.default_rng(42)
X = rng.standard_normal((n_rows, n_cols))


# ==============================================================================
# 2. Define a temporary file path
# ==============================================================================

fd, bin_file = tempfile.mkstemp(suffix=".bin")
os.close(fd)   # release OS file descriptor; numpy_to_memmap will re-open the path
print(f"2. Temporary file for memory mapping: {bin_file}\n")


# ==============================================================================
# Main demo — sections 3-8 wrapped in try/finally for safe cleanup
# ==============================================================================

try:
    # ==========================================================================
    # 3. Convert to Memory-Mapped Matrix
    # ==========================================================================

    print("3. Converting the in-memory matrix to a MemoryMappedMatrix ...")
    X_mmap = numpy_to_memmap(bin_file, X)
    print("   Done.\n")


    # ==========================================================================
    # 4. Explore the object
    # ==========================================================================

    print("4. Exploring the MemoryMappedMatrix object:")
    print(f"   rows()  = {X_mmap.rows()}")
    print(f"   cols()  = {X_mmap.cols()}")
    print(f"   size()  = {X_mmap.size()}   (rows * cols = {X_mmap.rows() * X_mmap.cols()})\n")


    # ==========================================================================
    # 5. Extract a block (subsetting via the numpy view)
    # ==========================================================================

    print("5. Extracting a 3x3 block via to_numpy()[0:3, 0:3]:")
    block = X_mmap.to_numpy()[0:3, 0:3]
    print(np.round(block, 4))
    match = np.allclose(block, X[0:3, 0:3])
    print(f"   Matches original matrix block? {match}\n")


    # ==========================================================================
    # 6. Read back into memory
    # ==========================================================================

    print("6. Reading the full memory-mapped matrix back into a NumPy array:")
    # .to_numpy() returns a zero-copy view; .copy() gives an independent array.
    X_restored = X_mmap.to_numpy().copy()
    match_full = np.allclose(X, X_restored)
    print(f"   Matches original full matrix? {match_full}\n")


    # ==========================================================================
    # 7. Re-open the file in read-only mode
    # ==========================================================================

    print("7. Re-opening the existing binary file in ReadOnly mode:")
    X_ro = MemoryMappedMatrix(bin_file, n_rows, n_cols, AccessMode.ReadOnly)
    print(f"   rows() = {X_ro.rows()},  cols() = {X_ro.cols()}\n")


    # ==========================================================================
    # 8. Out-of-core data generation: write column-by-column to a new file
    # ==========================================================================

    print("8. Out-of-core generation: writing data column-by-column to a new file.\n")
    print("   For large datasets, streaming data column-by-column to disk is")
    print("   significantly more memory-efficient than building the entire array in")
    print("   RAM first. Element-wise access via mmap[i, j] is demonstrated in")
    print("   Section 9 below.\n")

    fd_ooc, bin_file_ooc = tempfile.mkstemp(suffix=".bin")
    os.close(fd_ooc)

    try:
        rng_ooc = np.random.default_rng(0)
        print("   Writing data column-by-column to stream it efficiently ...")
        with open(bin_file_ooc, "wb") as fh:
            for _ in range(n_cols):
                # Each column is written as a contiguous float64 block.
                # Column-major (Fortran) order matches C++ Eigen ColMajor storage.
                col = rng_ooc.standard_normal(n_rows).astype(np.float64)
                fh.write(col.tobytes())

        print("   Mapping the sequentially populated binary file ...")
        X_ooc = MemoryMappedMatrix(bin_file_ooc, n_rows, n_cols, AccessMode.ReadOnly)
        print(f"   rows() = {X_ooc.rows()},  cols() = {X_ooc.cols()}")
        print(f"   First element: {X_ooc[0, 0]:.6f}   (read via __getitem__ — to_numpy() requires ReadWrite mode)\n")
    finally:
        try:
            os.unlink(bin_file_ooc)
            print("   -> Out-of-core temporary file cleaned up.")
        except OSError:
            pass


    # ==========================================================================
    # 9. Element-wise access: __getitem__ and __setitem__
    # ==========================================================================

    print("9. Element-wise access via mmap[row, col] (0-based indices).\n")

    # --- 9a. Scalar read ---
    r, c = 2, 3
    val = X_mmap[r, c]
    assert isinstance(val, float), "Expected a plain float scalar"
    match_scalar = np.isclose(val, X_mmap.to_numpy()[r, c])
    print(f"   Read  X_mmap[{r}, {c}] = {val:.6f}   (matches numpy view: {match_scalar})")

    # --- 9b. Scalar write ---
    original_val = X_mmap[r, c]
    X_mmap[r, c] = 99.0
    readback = X_mmap[r, c]
    numpy_readback = X_mmap.to_numpy()[r, c]
    print(f"   Write X_mmap[{r}, {c}] = 99.0 -> read back = {readback:.1f}   "
          f"(numpy view agrees: {np.isclose(numpy_readback, 99.0)})")

    # Revert to the original value so the rest of the demo is unaffected
    X_mmap[r, c] = original_val
    print(f"   Reverted to original value: {X_mmap[r, c]:.6f}")

    # --- 9c. Bounds guard ---
    print(f"   Bounds check (row index {n_rows} on a {n_rows}-row matrix):")
    try:
        _ = X_mmap[n_rows, 0]   # out-of-bounds
    except IndexError as e:
        print(f"   Caught IndexError: {e}")

    # --- 9d. Mode guard ---
    print(f"   Write attempt on ReadOnly matrix:")
    try:
        X_ro[0, 0] = 1.0        # X_ro was opened ReadOnly in Section 7
    except RuntimeError as e:
        print(f"   Caught RuntimeError: {e}")

    # --- 9e. Block write via slice __setitem__ ---
    block_rows = slice(0, 3)
    block_cols = slice(0, 4)
    new_block = np.arange(12.0, dtype=np.float64).reshape(3, 4)
    original_block = X_mmap.to_numpy()[block_rows, block_cols].copy()
    X_mmap[block_rows, block_cols] = new_block
    readback_block = X_mmap.to_numpy()[block_rows, block_cols]
    block_match = np.allclose(readback_block, new_block)
    print(f"   Block write X_mmap[0:3, 0:4] = arange(12).reshape(3,4)  "
          f"-> round-trip match: {block_match}")

    # Revert block
    X_mmap[block_rows, block_cols] = original_block
    print(f"   Reverted block.")

    # --- 9f. Explicit write_block() method ---
    new_block2 = np.full((3, 4), fill_value=7.0, dtype=np.float64)
    original_block2 = X_mmap.to_numpy()[0:3, 0:4].copy()
    X_mmap.write_block(0, 3, 0, 4, new_block2)
    readback_block2 = X_mmap.to_numpy()[0:3, 0:4]
    block2_match = np.allclose(readback_block2, 7.0)
    print(f"   write_block(0,3, 0,4, fill=7.0)            "
          f"-> round-trip match: {block2_match}")

    # Revert
    X_mmap.write_block(0, 3, 0, 4, original_block2)
    print(f"   Reverted block.")

    print()

finally:
    try:
        os.unlink(bin_file)
        print("   -> Temporary file cleaned up.")
    except OSError:
        pass

print("\nDemo completed successfully!")

File: R/memory_mapping/demo_memory_mapping.R
# ==============================================================================
# demo_memory_mapping.R
# ==============================================================================
#
# This script demonstrates the memory-mapping functionalities for the
# TRexSelector package.
# Memory-mapped matrices allow handling large datasets without loading
# the entire matrix into R's active memory, seamlessly passing a pointer to the
# C++ backend.
#
# Note: File handling in this demo
# Memory-mapped matrices are written to a temporary binary in a temporary
# directory. The file is automatically cleaned up when the R session ends.
#
# ==============================================================================

library(TRexSelector)

cat("=================================================================\n")
cat(" Memory Mapping Demo for TRexSelector\n")
cat("=================================================================\n\n")

# 1. Create a sample in-memory matrix
n_rows <- 100
n_cols <- 50
cat(sprintf("1. Creating a sample %d x %d numeric matrix in R.\n", n_rows, n_cols))
set.seed(42)
mat <- matrix(rnorm(n_rows * n_cols), nrow = n_rows, ncol = n_cols)


# 2. Define a temporary file path
bin_file <- tempfile(pattern = "trex_data_", fileext = ".bin")
cat(sprintf("2. Temporary file for memory mapping: %s\n", bin_file))

# Ensure cleanup on exit
on.exit({
  if (file.exists(bin_file)) {
    unlink(bin_file)
    cat("   -> Temporary file cleaned up.\n")
  }
}, add = TRUE)


# 3. Convert to Memory-Mapped Matrix
cat("\n3. Converting the in-memory matrix to a MemoryMappedMatrix...\n")
mmap_mat <- convert_to_memory_mapped(mat, bin_file)


# 4. Explore the object using S3 methods
cat("\n4. Exploring the mmap_matrix object:\n")
print(mmap_mat)

cat("\n   Dimensions of the mmap_matrix:\n")
print(dim(mmap_mat))


# 5. Extract blocks (subsetting)
cat("\n5. Extracting a 3x3 block using the '[' operator:\n")
block <- mmap_mat[1:3, 1:3]
print(block)

# Verify against the original matrix
cat("   Matches original matrix block? ", isTRUE(all.equal(block, mat[1:3, 1:3])), "\n")


# 6. Read back into memory
cat("\n6. Reading the full memory-mapped matrix back into an R matrix:\n")
mat_restored <- as.matrix(mmap_mat)
cat("   Matches original full matrix? ", isTRUE(all.equal(mat, mat_restored)), "\n")


# 7. Re-open the file in read-only mode
cat("\n7. Re-opening the existing binary file in 'readonly' mode:\n")
mmap_readonly <- mmap_matrix(bin_file, rows = n_rows, cols = n_cols, mode = "readonly")
print(mmap_readonly)


# 8. Out-of-core data generation: Create file on disk and fill it incrementally
cat("\n8. Out-of-core generation: Creating file on disk and filling it incrementally:\n")
cat("   For large datasets, streaming data column-by-column to disk is significantly\n")
cat("   more memory-efficient than building the entire matrix in RAM first.\n")
cat("   Element-wise access via [ and [<- is demonstrated in Section 9 below.\n\n")

bin_file_ooc <- tempfile(pattern = "trex_data_ooc_", fileext = ".bin")

# Ensure cleanup for the second file
on.exit({
  if (file.exists(bin_file_ooc)) {
    unlink(bin_file_ooc)
    cat("   -> Out-of-core temporary file cleaned up.\n")
  }
}, add = TRUE)

cat("   Writing data column-by-column to stream it efficiently...\n")
con <- file(bin_file_ooc, "wb")
for (i in seq_len(n_cols)) {
  writeBin(as.double(rnorm(n_rows)), con)
}
close(con)

cat("   Mapping the sequentially populated binary file...\n")
mmap_ooc <- mmap_matrix(bin_file_ooc, rows = n_rows, cols = n_cols)
print(mmap_ooc)


# 9. Element-wise access: [ (read) and [<- (write)
cat(strrep("=", 65), "\n")
cat("9. Element-wise access via [ and [<- (1-based indices).\n\n")

# --- 9a. Scalar read returns a plain double, not a 1x1 matrix ---
r <- 2L
cc <- 3L
val <- mmap_mat[r, cc]
cat(sprintf("   Read  mmap_mat[%d, %d] = %.6f\n", r, cc, val))
cat(sprintf("   Is plain scalar (not matrix): %s\n", !is.matrix(val)))
cat(sprintf("   Matches original mat[%d, %d]? %s\n\n", r, cc,
            isTRUE(all.equal(val, mat[r, cc]))))

# --- 9b. Scalar write ---
orig_val <- mmap_mat[r, cc]
mmap_mat[r, cc] <- 99.0
readback <- mmap_mat[r, cc]
cat(sprintf("   Write mmap_mat[%d, %d] <- 99.0  -> read back = %.1f   (correct: %s)\n\n",
            r, cc, readback, isTRUE(all.equal(readback, 99.0))))

# Revert
mmap_mat[r, cc] <- orig_val
cat(sprintf("   Reverted to original value: %.6f\n\n", mmap_mat[r, cc]))

# --- 9c. Block write ---
block_rows <- 1:3
block_cols <- 1:4
new_block  <- matrix(seq_len(length(block_rows) * length(block_cols)), nrow = length(block_rows))
mmap_mat[block_rows, block_cols] <- new_block
readback_block <- mmap_mat[block_rows, block_cols]
cat(sprintf("   Block write mmap_mat[1:3, 1:4] <- matrix(1:12):\n"))
cat(sprintf("   Round-trip matches? %s\n\n", isTRUE(all.equal(readback_block, new_block))))

# Revert block to original
mmap_mat[block_rows, block_cols] <- mat[block_rows, block_cols]

# --- 9d. Mode guard ---
cat("   Write attempt on ReadOnly matrix:\n")
tryCatch(
  mmap_readonly[1, 1] <- 0,
  error = function(e) cat(sprintf("   Caught error: %s\n", conditionMessage(e)))
)


cat("\nDemo completed successfully!\n")
